File: backend/core/routers/data_enrichment.py


# [AI GENERATED]
# Model: GitHub Copilot (GPT-4.1)
# Logic: Scaffold a FastAPI router for Data Enrichment Pipelines, providing endpoints to register, run, and monitor enrichment jobs integrating external data sources.
# Why: Roadmap item 24 requires integration of external data sources for data enrichment.
# Root Cause: No API existed for managing or executing enrichment pipelines.
# Context: This router provides endpoints for registering pipelines, running jobs, and checking job status. Future: integrate with persistent storage, add scheduling, and support custom connectors. For advanced enrichment orchestration, consider using a more advanced model (Claude Opus).

from fastapi import APIRouter, status, HTTPException
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime

router = APIRouter()

# In-memory store for demo purposes
ENRICHMENT_PIPELINES = []
ENRICHMENT_JOBS = []

class EnrichmentPipeline(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    source: str  # e.g., 'external_api', 'csv_upload'
    target_dataset: str
    created_at: datetime
    config: Optional[dict] = None

class EnrichmentJob(BaseModel):
    id: int
    pipeline_id: int
    status: str  # e.g., 'pending', 'running', 'completed', 'failed'
    started_at: datetime
    finished_at: Optional[datetime] = None
    result: Optional[str] = None

@router.post("/data-enrichment/pipelines", status_code=status.HTTP_201_CREATED)
def register_pipeline(pipeline: EnrichmentPipeline):
    if any(p.id == pipeline.id for p in ENRICHMENT_PIPELINES):
        raise HTTPException(status_code=400, detail="Pipeline ID already exists")
    ENRICHMENT_PIPELINES.append(pipeline)
    return {"message": "Pipeline registered", "pipeline_id": pipeline.id}

@router.get("/data-enrichment/pipelines", response_model=List[EnrichmentPipeline])
def list_pipelines():
    return ENRICHMENT_PIPELINES

@router.post("/data-enrichment/jobs", status_code=status.HTTP_201_CREATED)
def run_enrichment_job(job: EnrichmentJob):
    ENRICHMENT_JOBS.append(job)
    return {"message": "Job started", "job_id": job.id}

@router.get("/data-enrichment/jobs", response_model=List[EnrichmentJob])
def list_jobs(pipeline_id: Optional[int] = None):
    if pipeline_id:
        return [j for j in ENRICHMENT_JOBS if j.pipeline_id == pipeline_id]
    return ENRICHMENT_JOBS

File: services/frontend/src/services/dataEnrichment.js
import axios from 'axios';

const API_BASE = '/api';

export const registerPipeline = async (pipeline) => {
    const res = await axios.post(`${API_BASE}/data-enrichment/pipelines`, pipeline);
    return res.data;
};

export const listPipelines = async () => {
    const res = await axios.get(`${API_BASE}/data-enrichment/pipelines`);
    return res.data;
};

export const runEnrichmentJob = async (job) => {
    const res = await axios.post(`${API_BASE}/data-enrichment/jobs`, job);
    return res.data;
};

export const listJobs = async (pipelineId) => {
    const url = pipelineId ? `${API_BASE}/data-enrichment/jobs?pipeline_id=${pipelineId}` : `${API_BASE}/data-enrichment/jobs`;
    const res = await axios.get(url);
    return res.data;
};
